storage/streams/src/main/java/io/apicurio/registry/streams/StreamsRegistryStorage.java

```diff
@@ -0,0 +1,420 @@
+    @Override
+    public StoredArtifact getArtifact(String artifactId) throws ArtifactNotFoundException, RegistryStorageException {
+        return addContent(getLastArtifact(artifactId));
+    }
+
+    @Override
+    public ArtifactMetaDataDto updateArtifact(String artifactId, ArtifactType artifactType, String content) throws ArtifactNotFoundException, RegistryStorageException {
+        Str.Data data = storageStore.get(artifactId);
+        if (data != null) {
+            if (data.getArtifactsCount() == 0) {
+                throw new ArtifactNotFoundException(artifactId);
+            }
+        }
+
+        CompletableFuture<RecordMetadata> submitCF = submitter.submitArtifact(Str.ActionType.UPDATE, artifactId, -1, artifactType, content);
+        RecordMetadata rmd = ConcurrentUtil.get(submitCF);
+        long offset = rmd.offset();
+        int partition = rmd.partition();
+        CompletionStage<Str.Data> dataCF = storageFunction.apply(artifactId, offset);
+        data = ConcurrentUtil.result(dataCF);
+        long globalId = properties.toGlobalId(offset, partition);
+        for (int i = data.getArtifactsCount() - 1; i >=0; i--) {
+            Str.ArtifactValue value = data.getArtifacts(i);
+            if (value.getId() == globalId) {
+                return MetaDataKeys.toArtifactMetaData(value.getMetadataMap());
+            }
+        }
+        throw new ArtifactNotFoundException(artifactId);
+    }
+
+    @Override
+    public Set<String> getArtifactIds() {
+        Set<String> ids = new TreeSet<>();
+        try (CloseableIterator<String> iter = storageStore.allKeys()) {
+            while (iter.hasNext()) {
+                ids.add(iter.next());
+            }
+            ids.remove(GLOBAL_RULES_ID);
+            return ids;
+        }
+    }
+
+    @Override
+    public ArtifactMetaDataDto getArtifactMetaData(String artifactId) throws ArtifactNotFoundException, RegistryStorageException {
+        Map<String, String> content = getLastArtifact(artifactId).getMetadataMap();
+        return MetaDataKeys.toArtifactMetaData(content);
+    }
+
+    @Override
+    public void updateArtifactMetaData(String artifactId, EditableArtifactMetaDataDto metaData) throws ArtifactNotFoundException, RegistryStorageException {
+        Str.Data data = storageStore.get(artifactId);
+        if (data != null) {
+            ConcurrentUtil.get(submitter.submitMetadata(Str.ActionType.UPDATE, artifactId, -1, metaData.getName(), metaData.getDescription()));
+        } else {
+            throw new ArtifactNotFoundException(artifactId);
+        }
+    }
+
+    @Override
+    public List<RuleType> getArtifactRules(String artifactId) throws ArtifactNotFoundException, RegistryStorageException {
+        Str.Data data = storageStore.get(artifactId);
+        if (data != null) {
+            return data.getRulesList().stream().map(v -> RuleType.fromValue(v.getType().name())).collect(Collectors.toList());
+        }
+        if (isGlobalRules(artifactId)) {
```

Comment:
More efficient to put this branch before the `stateStore.get()`, or is this global rules thing a global _default_?

Reply:
I'm reusing (or abusing) Data also to store global rules.
If no Data is found for global rules, it's OK to return empty collection (to be in-sync with other impls).

Reply:
So if global rules are semantically different from real data (i.e. globals rules are not functioning as a default) then it would be better to put the globals rules check (a simple `equals()`) before the more expensive call to `storageStore.get()`.

Reply:
Global rules are Data -- just one without any artifacts.
This way I don't have to keep any additional state or handle global rules any different -- apart from this funky, but minor, checks in the impl itself.
I still need to do storageStore::get in order to get this global rules Data -- to see if it has any rules values.